example/particles/particles.cpp

```diff
@@ -0,0 +1,361 @@
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include "particles.hpp"
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+// *************************************************//
+// redefine some internal parthenon functions      *//
+// *************************************************//
+namespace particles_example {
+
+Packages_t ProcessPackages(std::unique_ptr<ParameterInput> &pin) {
+  Packages_t packages;
+  packages["particles_package"] = particles_example::Particles::Initialize(pin.get());
+  return packages;
+}
+
+void ProblemGenerator(MeshBlock *pmb, ParameterInput *pin) {
+  // Don't do anything for now
+}
+
+// *************************************************//
+// define the "physics" package particles_package, *//
+// which includes defining various functions that  *//
+// control how parthenon functions and any tasks   *//
+// needed to implement the "physics"               *//
+// *************************************************//
+
+namespace Particles {
+
+std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
+  auto pkg = std::make_shared<StateDescriptor>("particles_package");
+
+  int num_particles = pin->GetOrAddInteger("Particles", "num_particles", 100);
+  pkg->AddParam<>("num_particles", num_particles);
+  Real particle_speed = pin->GetOrAddReal("Particles", "particle_speed", 1.0);
+  pkg->AddParam<>("particle_speed", particle_speed);
+  Real const_dt = pin->GetOrAddReal("Particles", "const_dt", 1.0);
+  pkg->AddParam<>("const_dt", const_dt);
+
+  // Initialize random number generator pool
+  int rng_seed = pin->GetInteger("Particles", "rng_seed");
+  pkg->AddParam<>("rng_seed", rng_seed);
+  RNGPool rng_pool(rng_seed);
+  pkg->AddParam<>("rng_pool", rng_pool);
+
+  std::string swarm_name = "my particles";
+  Metadata swarm_metadata;
+  pkg->AddSwarm(swarm_name, swarm_metadata);
+  Metadata real_swarmvalue_metadata({Metadata::Real});
+  pkg->AddSwarmValue("vx", swarm_name, real_swarmvalue_metadata);
+  pkg->AddSwarmValue("vy", swarm_name, real_swarmvalue_metadata);
+  pkg->AddSwarmValue("vz", swarm_name, real_swarmvalue_metadata);
+  pkg->AddSwarmValue("weight", swarm_name, real_swarmvalue_metadata);
+
+  std::string field_name = "particle_deposition";
+  Metadata m({Metadata::Cell, Metadata::Independent});
+  pkg->AddField(field_name, m);
+
+  pkg->EstimateTimestep = EstimateTimestep;
+
+  return pkg;
+}
+
+AmrTag CheckRefinement(Container<Real> &rc) { return AmrTag::same; }
+
+Real EstimateTimestep(std::shared_ptr<Container<Real>> &rc) {
```

Comment:
I would recommend that you read through section R.37 here https://www.modernescpp.com/index.php/c-core-guidelines-passing-smart-pointer. 

Instead of passing the shared pointer as a reference I believe it is better practice to dereference the Container and pass it in. Your function should look like this:

```
Real EstimateTimestep(Container<Real> &rc) {
```

I believe you only need to pass a shared pointer into a function if you are going to be sharing ownership of the object. Let me know if this makes sense it was a little confusing to me at first.

Reply:
Alternatively, he could pass in the raw pointer, no?

Reply:
True, I guess it's a style choice. This is from the google style guide, they recommend using pointers if you are going to be changing the object (it's non const), which is inline with what your saying @Yurlungur :

> Reference Arguments
>
> All parameters passed by lvalue reference must be labeled const.
>
> In C, if a function needs to modify a variable, the parameter must use a pointer, eg int foo(int *pval). In C++, the function can alternatively declare a reference parameter: int foo(int &val).
>
>Defining a parameter as reference avoids ugly code like (*pval)++. Necessary for some applications like copy constructors. Makes it clear, unlike with pointers, that a null pointer is not a possible value.
>
> References can be confusing, as they have value syntax but pointer semantics.
> 
> Within function parameter lists all references must be const:
> 
> void Foo(const string &in, string *out);
> 
> In fact it is a very strong convention in Google code that input arguments are values or const references while output arguments are pointers. See the Pass by Value section for the cases where we allow pass-by-value. Input parameters may be const pointers, but we never allow non-const reference parameters except when required by convention, e.g., swap().
> 
> However, there are some instances where using const T* is preferable to const T& for input parameters. For example:
> 
>     You want to pass in a null pointer.
>     The function saves a pointer or reference to the input.
> 
> Remember that most of the time input parameters are going to be specified as const T&. Using const T* instead communicates to the reader that the input is somehow treated differently. So if you choose const T* rather than const T&, do so for a concrete reason; otherwise it will likely confuse readers by making them look for an explanation that doesn't exist.

Reply:
We've used raw pointers for this elsewhere in the code, so I don't think it's unreasonable to use those here. I agree, it comes down to stylistic preference.

Reply:
This is a broader issue than just this example, right? (It also shows up in the advection example, for example, and in interface/update.hpp the declaration is `Real EstimateTimestep(std::shared_ptr<Container<Real>> &rc)`. Maybe there needs to be a significant review of how `shared_ptr`s are used in the code.

Reply:
That's find go ahead and create an issue. This can also be addressed later. This is not a breaking issue. Consistency is probably more important at this point.

Reply:
See #316